sklearn/covariance/_shrunk_covariance.py

```diff
@@ -106,12 +106,13 @@ def shrunk_covariance(emp_cov, shrinkage=0.1):
 
     where `mu = trace(cov) / n_features`.
     """
-    emp_cov = check_array(emp_cov)
-    n_features = emp_cov.shape[0]
+    emp_cov = check_array(emp_cov, allow_nd=True)
+    n_features = emp_cov.shape[-1]
 
-    mu = np.trace(emp_cov) / n_features
     shrunk_cov = (1.0 - shrinkage) * emp_cov
-    shrunk_cov.flat[:: n_features + 1] += shrinkage * mu
+    mu = np.trace(emp_cov, axis1=-2, axis2=-1) / n_features
+    mu = np.expand_dims(mu, axis=tuple(range(mu.ndim, emp_cov.ndim)))
+    shrunk_cov += shrinkage * mu * np.eye(n_features)
```

Comment:
Here, we materialize the `eye` matrix. If we restrain to a 3-D matrix then, we can efficiently do the in place addition with flattening as previously done.